File: android/app/src/main/kotlin/com/example/access_checkout_flutter_demo/AccessCheckoutView.kt
package com.example.access_checkout_flutter_demo

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.view.LayoutInflater
import android.view.View
import androidx.lifecycle.LifecycleOwner
import com.worldpay.access.checkout.client.api.exception.AccessCheckoutException
import com.worldpay.access.checkout.client.session.AccessCheckoutClient
import com.worldpay.access.checkout.client.session.AccessCheckoutClientBuilder
import com.worldpay.access.checkout.client.session.listener.SessionResponseListener
import com.worldpay.access.checkout.client.session.model.CardDetails
import com.worldpay.access.checkout.client.session.model.SessionType
import com.worldpay.access.checkout.client.validation.AccessCheckoutValidationInitialiser
import com.worldpay.access.checkout.client.validation.config.CardValidationConfig
import com.worldpay.access.checkout.client.validation.listener.AccessCheckoutCardValidationListener
import com.worldpay.access.checkout.client.validation.model.CardBrand
import com.worldpay.access.checkout.ui.AccessCheckoutEditText
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView


class AccessCheckoutView(
    private val lifecycleOwner: LifecycleOwner,
    private val methodChannel: MethodChannel,
    context: Context,
    creationParams: Map<String, Any>,
) :
    PlatformView {
    private val layout: View
    private val panInput: AccessCheckoutEditText
    private val expiryInput: AccessCheckoutEditText
    private val cvcInput: AccessCheckoutEditText

    private var baseUrl: String = creationParams["baseUrl"] as String
    private var checkoutId: String = creationParams["checkoutId"] as String
    private var useCardValidation: Boolean =
        creationParams["useCardValidation"] as? Boolean ?: false

    private var accessCheckoutClient: AccessCheckoutClient

    init {
        layout = LayoutInflater.from(context)
            .inflate(R.layout.access_checkout_layout, null)

        panInput = layout.findViewById(R.id.pan_input)
        expiryInput = layout.findViewById(R.id.expiry_date_input)
        cvcInput = layout.findViewById(R.id.cvc_input)

        if (useCardValidation) {
            initializeCardValidation()
        }
        accessCheckoutClient = AccessCheckoutClientBuilder()
            .baseUrl(baseUrl)
            .checkoutId(checkoutId)
            .lifecycleOwner(lifecycleOwner)
            .sessionResponseListener(
                object : SessionResponseListener {
                    override fun onError(error: AccessCheckoutException) {
                        methodChannel.invokeMethod("onSessionError", "Could not create session")
                    }

                    override fun onSuccess(sessionResponseMap: Map<SessionType, String>) {
                        // Important: Flutter will not understand SessionType
                        // therefore it needs to be converted into a JSON-serializable format
                        val sessionData = sessionResponseMap.mapKeys { it.key.name }
                        methodChannel.invokeMethod("onSessionGenerated", sessionData)
                    }
                }
            )
            .context(context)
            .build()

        methodChannel.setMethodCallHandler { call, result ->
            when (call.method) {
                "generateSession" -> generateSession()
                else -> result.notImplemented()
            }
        }
    }

    //    private fun styleComponent() {}
    private fun initializeCardValidation() {
        val cardValidationConfig = CardValidationConfig.Builder()
            .baseUrl(baseUrl)
            .pan(panInput)
            .expiryDate(expiryInput)
            .cvc(cvcInput)
            .lifecycleOwner(lifecycleOwner)
            .enablePanFormatting()
            .validationListener(object : AccessCheckoutCardValidationListener {
                override fun onBrandChange(cardBrand: CardBrand?) {
                    // TODO: Update the brand image using your SVG loader
                }

                override fun onCvcValidated(isValid: Boolean) {
                    updateUIField(cvcInput, isValid)
                    if (!isValid) {
                        methodChannel.invokeMethod("onValidationUpdated", false)
                    }
                }

                override fun onExpiryDateValidated(isValid: Boolean) {
                    updateUIField(expiryInput, isValid)
                    if (!isValid) {
                        methodChannel.invokeMethod("onValidationUpdated", false)
                    }
                }

                override fun onPanValidated(isValid: Boolean) {
                    updateUIField(panInput, isValid)
                    if (!isValid) {
                        methodChannel.invokeMethod("onValidationUpdated", false)
                    }

                }

                override fun onValidationSuccess() {
                    methodChannel.invokeMethod("onValidationUpdated", true)
                }

            })
            .build()

        AccessCheckoutValidationInitialiser.initialise(cardValidationConfig)
    }

    private fun updateUIField(field: AccessCheckoutEditText, isValid: Boolean) {
        val colour = if (isValid) Color.GREEN else Color.RED

        //Update Text colour
        field.setTextColor(colour)
        //Update Border
        val border = GradientDrawable()
        border.setStroke(3, colour)
        field.background = border
    }


    private fun generateSession() {

        val cardDetails = CardDetails.Builder()
            .pan(panInput)
            .expiryDate(expiryInput)
            .cvc(cvcInput)
            .build()


        accessCheckoutClient.generateSessions(cardDetails, listOf(SessionType.CARD))
    }

    override fun getView(): View = layout
    override fun dispose() {}
}

File: ios/Runner/AccessCheckoutView.swift
import AccessCheckoutSDK
import Flutter
import UIKit

class AccessCheckoutView: NSObject, FlutterPlatformView {
    @IBOutlet private var panInput: AccessCheckoutUITextField!
    @IBOutlet private var expiryInput: AccessCheckoutUITextField!
    @IBOutlet private var cvcInput: AccessCheckoutUITextField!
    private var accessCheckoutClient: AccessCheckoutClient
    private var methodChannel: FlutterMethodChannel
    private var _view: UIView

    private var baseUrl: String
    private var checkoutId: String
    private var useCardValidation: Bool
    
    private var controller: AccessCheckoutViewController


    init(
        methodChannel channel: FlutterMethodChannel,
        frame: CGRect,
        viewIdentifier viewId: Int64,
        binaryMessenger messenger: FlutterBinaryMessenger?,
        creationParams: [String: Any]
    ) {
        baseUrl = creationParams["baseUrl"] as? String ?? ""
        checkoutId = creationParams["checkoutId"] as? String ?? ""
        useCardValidation = creationParams["useCardValidation"] as? Bool ?? false
        methodChannel = channel

        accessCheckoutClient = try! AccessCheckoutClientBuilder()
            .accessBaseUrl(baseUrl)
            .checkoutId(checkoutId)
            .build()
        
        let storyboard = UIStoryboard(name: "AccessCheckoutView", bundle: Bundle.main)
        controller = storyboard.instantiateViewController(withIdentifier:"ViewController") as! AccessCheckoutViewController
     
        _view = controller.view!
      
        super.init()

        methodChannel.setMethodCallHandler({
            [weak self] (call, result) in
            guard let self = self else {return}

            switch call.method{
            case "generateSession":
                generateSession()
            default:
                result(FlutterMethodNotImplemented)

            }
        })

        // iOS views can be created here
        referenceNativeView()

        if(useCardValidation){
            initializeCardValidation()
        }
    }

    func initializeCardValidation() {
        let validationConfig = try! CardValidationConfig.builder()
            .pan(panInput)
            .expiryDate(expiryInput)
            .cvc(cvcInput)
            .accessBaseUrl(baseUrl)
            .validationDelegate(self)
            .enablePanFormatting()
                .build()


        AccessCheckoutValidationInitialiser().initialise(validationConfig)
    }

    func updateUIField(field: AccessCheckoutUITextField, isValid: Bool) {
        var  colour = isValid ? UIColor.green : UIColor.red

        //Update Text & border colour
        field.textColor = colour
        field.borderColor = colour
    }

    func generateSession() {

        let cardDetails = try! CardDetailsBuilder().pan(panInput)
            .expiryDate(expiryInput)
            .cvc(cvcInput)
            .build()

        do { try accessCheckoutClient.generateSessions(
            cardDetails: cardDetails,
            sessionTypes: [SessionType.card]) { result in
                DispatchQueue.main.async {
                    switch result {
                    case .failure(let error):
                        print(error)
                        self.methodChannel.invokeMethod("onSessionError", arguments: "Could not create session")

                    case .success(let sessions):
                        let sessionData = sessions.mapValues{$0}
                        self.methodChannel.invokeMethod("onSessionGenerated", arguments: sessionData)
                    }
                }
            }
        }
        catch {
            self.methodChannel.invokeMethod("onSessionError", arguments: "Could not create session")
        }
    }

    func referenceNativeView() {
        panInput = controller.panInput
        expiryInput = controller.expiryInput
        cvcInput = controller.cvcInput
    }


    func view() -> UIView {
        return _view
    }

}

extension AccessCheckoutView: AccessCheckoutCardValidationDelegate {
    func cardBrandChanged(cardBrand: AccessCheckoutSDK.CardBrand?) {
        //TODO
    }

    func panValidChanged(isValid: Bool) {
        self.updateUIField(field: self.panInput, isValid: isValid)
        if(!isValid){
            self.methodChannel.invokeMethod("OnValidationUpdated", arguments:false)
        }
    }

    func expiryDateValidChanged(isValid: Bool) {
        self.updateUIField(field: self.expiryInput, isValid: isValid)
        if(!isValid){
            self.methodChannel.invokeMethod("OnValidationUpdated", arguments:false)
        }
    }

    func cvcValidChanged(isValid: Bool) {
        self.updateUIField(field: self.cvcInput, isValid: isValid)
        if(!isValid){
            self.methodChannel.invokeMethod("OnValidationUpdated", arguments:false)
        }
    }

    func validationSuccess() {
        self.methodChannel.invokeMethod("OnValidationUpdated", arguments:true)

    }

}
